base/MAPL_NetCDF.F90

```diff
@@ -18,282 +14,247 @@ module MAPL_NetCDF
-   public :: is_time_unit
-   public :: is_valid_netcdf_datetime_string
-   public :: get_shift_sign
-   public :: split
-   public :: split_all
-   public :: lr_trim
+
+   ! LOW-LEVEL - keep commented out (private) unless debugging these procedures
+   public :: convert_NetCDF_DateTimeString_to_ESMF_Time, make_NetCDF_DateTime_duration, is_digit_string, get_shift_sign
+   public :: make_ESMF_TimeInterval, is_valid_netcdf_datetime_string, convert_to_integer, convert_to_real, split
+
+   interface make_ESMF_TimeInterval
+      module procedure :: make_ESMF_TimeInterval_integer
+      module procedure :: make_ESMF_TimeInterval_real
+   end interface make_ESMF_TimeInterval
+
+   interface make_NetCDF_DateTime_duration
+      module procedure :: make_NetCDF_DateTime_duration_integer
+      module procedure :: make_NetCDF_DateTime_duration_real
+   end interface make_NetCDF_DateTime_duration
+   
+   interface split
+      module procedure :: split_characters
+   end interface split
 
    character, parameter :: PART_DELIM = ' '
-   character, parameter :: ISO_DELIM = 'T'
    character, parameter :: DATE_DELIM = '-'
    character, parameter :: TIME_DELIM = ':'
+   character, parameter :: DELIMS(3) = [PART_DELIM, DATE_DELIM, TIME_DELIM]
+   character, parameter :: POINT = '.'
    character(len=*), parameter :: NETCDF_DATE = '0000' // DATE_DELIM // '00' // DATE_DELIM // '00'
    character(len=*), parameter :: NETCDF_TIME = '00' // TIME_DELIM // '00' // TIME_DELIM // '00'
-   character(len=*), parameter :: NETCDF_DATETIME_FORMAT = NETCDF_DATE // PART_DELIM // NETCDF_TIME
+   character(len=*), parameter :: NETCDF_DATETIME = NETCDF_DATE // PART_DELIM // NETCDF_TIME
    integer, parameter :: LEN_DATE = len(NETCDF_DATE)
    integer, parameter :: LEN_TIME = len(NETCDF_TIME)
-   integer, parameter :: LEN_NETCDF_DATETIME = len(NETCDF_DATETIME_FORMAT)
+   integer, parameter :: LEN_DATETIME = len(NETCDF_DATETIME)
+   integer, parameter :: NUM_PARTS_UNITS_STRING = 4
    character(len=*), parameter :: TIME_UNITS(7) = &
       [  'years       ', 'months      ', 'days        ', &
          'hours       ', 'minutes     ', 'seconds     ', 'milliseconds'    ]
    character, parameter :: SPACE = ' '
    type(ESMF_CalKind_Flag), parameter :: CALKIND_FLAG = ESMF_CALKIND_GREGORIAN
-   integer, parameter :: MAX_WIDTH = 10
+   character(len=*), parameter :: DIGIT_CHARS = '0123456789'
+   character, parameter :: PLUS = '+'
+   character, parameter :: MINUS = '-'
+   character(len=*), parameter :: SIGNS = PLUS // MINUS
+   character(len=*), parameter :: EMPTY_STRING = ''
+   integer, parameter :: MAX_CHARACTER_LENGTH = 64
 
 contains
 
-   ! Convert NetCDF_DateTime {int_time, units_string} to
-   ! ESMF time variables {interval, time0, time1} and time unit {tunit}
-   ! time0 is the start time, and time1 is time0 + interval
-   subroutine convert_NetCDF_DateTime_to_ESMF(int_time, units_string, &
-         interval, time0, unusable, time1, tunit, rc)
```

Comment:
Sorry @darianboggs, I gave you the wrong feedback about this subroutine.  Your first implementation is very advanced. It was ahead of me.   I am now using each of the options from your deleted line:

  subroutine convert_NetCDF_DateTime_to_ESMF(int_time, units_string, &
         interval, time0, unusable, time1, tunit, rc)
E.g. https://github.com/GEOS-ESM/MAPL/blob/feature/ygyu/traj_v2/gridcomps/History/MAPL_HistoryTrajectoryMod_smod.F90#L1212

Will you please keep those arguments (time0, time1)?   I find them very useful. 

When you try to build the code, you will find the MAPL_HistoryTrajectoryMod_smod.F90 needs to be changed.  You can change that to pass the build.

Reply:
OK. I'll keep the interface. The only questions are with the arguments. I have read more about CF units and `UDUNITS`. So I better understand what the `units` string represents, and I understand better why the duration can be integer or real of various kinds.

So my subroutine:
1) Converts the `units` string into a base or reference datetime (**`time0`**). That is specific to NetCDF CF Units. It creates an intermediate representation of the base datetime before creating the `ESMF_Time`.
2) Pulls out the datetime unit from the beginning of `units` (**`tunit`**).
3) Converts the real or integer duration since the base datetime and `tunit` into an `ESMF_TimeInterval` (**`interval`**).
4) Creates an `ESMF_Time` from `time0` and `interval` (**`time1`**`= time0 + interval`)

`time0` and `interval` are the only required outputs. `time1` and `tunit` are optional.
* Are `time0` and `time1` the best names for the two `ESMF_Time` variables?
* Should `optional` change for the four output arguments (`time0`, `interval`, `time1`, `tunit`)?

It will return the same arguments as before. It's just a matter of the names, positions, and `optional` attributes for the four output arguments.

From the MAPL/ESMF perspective, `time1` is the most important. From the NetCDF perspective, `tunit`, `interval`, and `time0` are more fundamental.

To convert the duration and `units` to `time1`, an `ESMF_Time`, the base datetime `time0` and `interval` are created first and added together. To convert `time1` to a NetCDF datetime, you need to know:
`tunits` *and*
* the original duration *or*
* `interval` *or*
* `time0`

Internally, I use a toolbox of four procedures to achieve 1-4.

I can leave it the way it is. I just wanted to check before we finalize it.

Thoughts?

Reply:
Hi @darianboggs : Thank you for your details description and questions.  Indeed it is matter of name convention and optional attribute for the interface. 

For the example Time =  2500 + "seconds since 1970-01-01 00:00:00"
I have a name option-2 for a reference (time_base may feel a bit strange, but overall it might be more literal).  Meanwhile I have been using your design and they work well for me already. Hence either way is good.


Name choice 1 | Name choice 2 | Values | Attribute
-- | -- | -- | --
Time0 | Time_base | 1970-01-01 00:00:00 | Optional
Interval | Time_increment | 2500 seconds | Optional
Time1 | Time_current | 2500 + "seconds since 1970-01-01 00:00:00" | Output
tunit | Unit_of_time | Seconds | Optional